File: src/main/java/me/rothes/protocolstringreplacer/packetlisteners/server/itemstack/WindowItems.java
package me.rothes.protocolstringreplacer.packetlisteners.server.itemstack;

import com.comphenix.protocol.PacketType;
import com.comphenix.protocol.events.PacketEvent;
import com.comphenix.protocol.wrappers.BukkitConverters;
import me.rothes.protocolstringreplacer.ProtocolStringReplacer;
import me.rothes.protocolstringreplacer.api.replacer.ReplacerConfig;
import me.rothes.protocolstringreplacer.api.user.PsrUser;
import me.rothes.protocolstringreplacer.replacer.ReplacerManager;
import org.bukkit.Material;
import org.bukkit.inventory.ItemStack;
import org.jetbrains.annotations.NotNull;

import java.util.List;

public final class WindowItems extends AbstractServerItemPacketListener {

    public WindowItems() {
        super(PacketType.Play.Server.WINDOW_ITEMS);
    }

    protected void process(@NotNull PacketEvent packetEvent) {
        PsrUser user = getEventUser(packetEvent);
        if (user == null) {
            return;
        }
        user.clearUserMetaCache();
        Object[] read = (Object[]) packetEvent.getPacket().getModifier().read(1);
        ReplacerManager replacerManager = ProtocolStringReplacer.getInstance().getReplacerManager();
        List<ReplacerConfig> nbt = replacerManager.getAcceptedReplacers(user, itemNbtFilter);
        List<ReplacerConfig> display = replacerManager.getAcceptedReplacers(user, itemDisplayFilter);
        List<ReplacerConfig> entries = replacerManager.getAcceptedReplacers(user, itemEntriesFilter);

        boolean saveMeta = !user.isInAnvil();
        for (Object item : read) {
            ItemStack itemStack = BukkitConverters.getItemStackConverter().getSpecific(item);
            if (itemStack.getType() == Material.AIR) {
                saveMeta = true;
                continue;
            }
            boolean blocked = replaceItemStack(packetEvent, user, listenType, itemStack, nbt, display, entries,
                    // Avoid too many packets kick
                    saveMeta);
            saveMeta = true;
            if (blocked) {
                return;
            }
        }
    }

}

File: bukkit/src/main/kotlin/io/github/rothes/protocolstringreplacer/packetlistener/server/itemstack/WindowItems.kt
package io.github.rothes.protocolstringreplacer.packetlistener.server.itemstack

import com.comphenix.protocol.PacketType
import com.comphenix.protocol.events.PacketEvent
import com.comphenix.protocol.wrappers.BukkitConverters
import io.github.rothes.protocolstringreplacer.ProtocolStringReplacer
import io.github.rothes.protocolstringreplacer.get
import io.github.rothes.protocolstringreplacer.set
import org.bukkit.Material

class WindowItems : BaseServerItemPacketListener(PacketType.Play.Server.WINDOW_ITEMS) {

    override fun process(packetEvent: PacketEvent) {
        val user = getEventUser(packetEvent) ?: return
        user.clearUserMetaCache()
        val replacerManager = ProtocolStringReplacer.getInstance().replacerManager
        val nbt = replacerManager.getAcceptedReplacers(user, itemNbtFilter)
        val lore = replacerManager.getAcceptedReplacers(user, itemLoreFilter)
        val entries = replacerManager.getAcceptedReplacers(user, itemEntriesFilter)

        var saveMeta = !user.isInAnvil
        packetEvent.packet.modifier[1] = (packetEvent.packet.modifier[1] as Array<*>).map {
            BukkitConverters.getItemStackConverter().getSpecific(it)
        }.map { itemStack ->
            if (itemStack.type == Material.AIR) {
                saveMeta = true
                return@map itemStack
            }
            replaceItemStack(packetEvent, user, listenType, itemStack, nbt, lore, entries, saveMeta).also {
                saveMeta = true
            } ?: return
        }.map {
            BukkitConverters.getItemStackConverter().getGeneric(it)
        }.toTypedArray()
    }
}
